packages/swap/src/routes/quoteV2.ts

```diff
@@ -0,0 +1,365 @@
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: bigint,
+  ): Promise<bigint>;
+  async function quoteLeg(
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: bigint | null,
+  ): Promise<bigint | null>;
+  async function quoteLeg(
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: bigint | null,
+  ): Promise<bigint | null> {
+    if (!amount) return null;
+
+    const id = randomUUID();
+
+    io.emit('quote_request', {
+      request_id: id,
+      amount: String(amount),
+      base_asset: getAssetAndChain(baseAsset),
+      quote_asset: getAssetAndChain(quoteAsset),
+    });
+
+    const quotes = await collectMakerQuotes(id, io.sockets.sockets.size, quotes$);
+
+    const { swappedAmount, remainingAmount } = await findPrice({
+      poolFee: 0,
+      amount: BigInt(amount),
+      limitOrders: quotes.flatMap(({ limit_orders: limitOrders }) =>
+        limitOrders.map(([tick, amt]) => ({ tick, amount: amt })),
+      ),
+    });
+
+    if (remainingAmount !== 0n) {
+      // TODO: handle remaining amount with the pool
+    }
+
+    return swappedAmount;
+  }
+
+  const getBestQuote = async (quoteRequest: ParsedQuoteParams & { amount: bigint }) => {
+    const { srcAsset, destAsset } = getInternalAssets(quoteRequest);
+    const networkFee = getPoolsNetworkFeeHundredthPips(env.CHAINFLIP_NETWORK);
+
+    const pools = await getPools(srcAsset, destAsset);
+
+    let intermediateAmount: bigint | null = null;
+    let outputAmount: bigint;
+    let networkFeeUsdc: bigint | null = null;
+    const firstLegPoolFee = getHundredthPipAmountFromAmount(
+      quoteRequest.amount,
+      pools[0].liquidityFeeHundredthPips,
+    );
+
+    const swapInputAmount = quoteRequest.amount - firstLegPoolFee;
+
+    const fees: SwapFee[] = [
+      {
+        type: 'LIQUIDITY',
+        amount: firstLegPoolFee.toString(),
+        ...getAssetAndChain(srcAsset),
+      },
+    ];
+
+    if (srcAsset !== 'Usdc' && destAsset !== 'Usdc') {
+      const leg1 = [srcAsset, 'Usdc'] as const;
+      const leg2 = ['Usdc', destAsset] as const;
+
+      let approximateUsdcAmount = await approximateIntermediateOutput(
+        srcAsset,
+        swapInputAmount.toString(),
+      );
+
+      networkFeeUsdc =
+        approximateUsdcAmount && getHundredthPipAmountFromAmount(approximateUsdcAmount, networkFee);
+
+      if (networkFeeUsdc && approximateUsdcAmount) {
+        approximateUsdcAmount -= networkFeeUsdc;
+      }
+
+      const quotes = await Promise.all([
+        quoteLeg(leg1, swapInputAmount),
+        quoteLeg(leg1, approximateUsdcAmount),
+      ]);
+
+      // TODO: check if approximated value is close enough or we need to get a second quote
+      if (quotes[1] === null) {
+        networkFeeUsdc = getHundredthPipAmountFromAmount(quotes[0], networkFee);
+        quotes[0] -= networkFeeUsdc;
+        quotes[1] = await quoteLeg(leg2, quotes[0]);
```

Comment:
this tries to quote the second leg a second time, right? what the reason for doing it already at the same time?

Reply:
The first time is an optimistic attempt to approximate the USDC value so we can quote both legs at once, rather than have to do them sequentially. if they are too far off, then we will quote it a second time